# Language Translation
In this project, you’re going to take a peek into the realm of neural network machine translation.  You’ll be training a sequence to sequence model on a dataset of English and French sentences that can translate new sentences from English to French.
## Get the Data
Since translating the whole language of English to French will take lots of time to train, we have provided you with a small portion of the English corpus.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import helper
import problem_unittests as tests

source_path = 'data/small_vocab_en'
target_path = 'data/small_vocab_fr'
source_text = helper.load_data(source_path)
target_text = helper.load_data(target_path)

## Explore the Data
Play around with view_sentence_range to view different parts of the data.

In [2]:
view_sentence_range = (0, 10)

"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np

print('Dataset Stats')
print('Roughly the number of unique words: {}'.format(len({word: None for word in source_text.split()})))

sentences = source_text.split('\n')
word_counts = [len(sentence.split()) for sentence in sentences]
print('Number of sentences: {}'.format(len(sentences)))
print('Average number of words in a sentence: {}'.format(np.average(word_counts)))

print()
print('English sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(source_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))
print()
print('French sentences {} to {}:'.format(*view_sentence_range))
print('\n'.join(target_text.split('\n')[view_sentence_range[0]:view_sentence_range[1]]))

Dataset Stats
Roughly the number of unique words: 227
Number of sentences: 137861
Average number of words in a sentence: 13.225277634719028

English sentences 0 to 10:
new jersey is sometimes quiet during autumn , and it is snowy in april .
the united states is usually chilly during july , and it is usually freezing in november .
california is usually quiet during march , and it is usually hot in june .
the united states is sometimes mild during june , and it is cold in september .
your least liked fruit is the grape , but my least liked is the apple .
his favorite fruit is the orange , but my favorite is the grape .
paris is relaxing during december , but it is usually chilly in july .
new jersey is busy during spring , and it is never hot in march .
our least liked fruit is the lemon , but my least liked is the grape .
the united states is sometimes busy during january , and it is sometimes warm in november .

French sentences 0 to 10:
new jersey est parfois calme pendant l' automne 

## Implement Preprocessing Function
### Text to Word Ids
As you did with other RNNs, you must turn the text into a number so the computer can understand it. In the function `text_to_ids()`, you'll turn `source_text` and `target_text` from words to ids.  However, you need to add the `<EOS>` word id at the end of each sentence from `target_text`.  This will help the neural network predict when the sentence should end.

You can get the `<EOS>` word id by doing:
```python
target_vocab_to_int['<EOS>']
```
You can get other word ids using `source_vocab_to_int` and `target_vocab_to_int`.

In [3]:
def text_to_ids(source_text, target_text, source_vocab_to_int, target_vocab_to_int):
    """
    Convert source and target text to proper word ids
    :param source_text: String that contains all the source text.
    :param target_text: String that contains all the target text.
    :param source_vocab_to_int: Dictionary to go from the source words to an id
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: A tuple of lists (source_id_text, target_id_text)
    """
    # Implement Function
#     source_id_text = []
#     for sentence in source_text.split('\n'):
#         source_sentence_id = [source_vocab_to_int[word] for word in sentence.split()]
#         source_id_text.append(source_sentence_id)
    
#     target_id_text = []
#     for sentence in target_text.split('\n'):
#         target_sentence_id = [target_vocab_to_int[word] for word in sentence.split()]
#         target_sentence_id.append(target_vocab_to_int['<EOS>'])
#         target_id_text.append(target_sentence_id)

    source_id_text = [[source_vocab_to_int[w] for w in sentence.split()] for sentence in source_text.split('\n')]
    target_id_text = [[target_vocab_to_int[w] for w in sentence.split()]+[target_vocab_to_int['<EOS>']] for sentence in target_text.split('\n')]   
    return source_id_text, target_id_text

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_text_to_ids(text_to_ids)

Tests Passed


### Preprocess all the data and save it
Running the code cell below will preprocess all the data and save it to file.

In [4]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
helper.preprocess_and_save_data(source_path, target_path, text_to_ids)

# Check Point
This is your first checkpoint. If you ever decide to come back to this notebook or have to restart the notebook, you can start from here. The preprocessed data has been saved to disk.

In [1]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import numpy as np
import helper

(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()

In [5]:
target_vocab_to_int

{',': 99,
 '-': 138,
 '-ce': 73,
 '-elle': 309,
 '-il': 52,
 '-ils': 263,
 '.': 203,
 '<EOS>': 1,
 '<GO>': 3,
 '<PAD>': 0,
 '<UNK>': 2,
 '?': 45,
 'a': 165,
 'agréable': 155,
 'aimait': 221,
 'aime': 119,
 'aiment': 187,
 'aiment-ils': 86,
 'aimeraient': 57,
 'aimez': 194,
 'aimons': 170,
 'aimé': 176,
 'aimé.': 265,
 'aimée': 11,
 'aimés': 270,
 'aller': 271,
 'allez': 127,
 'allions': 97,
 'allons': 331,
 'allé': 276,
 'allée': 178,
 'allés': 273,
 'amusant': 197,
 'anglais': 348,
 'animal': 323,
 'animaux': 5,
 'août': 279,
 'apprécié': 250,
 'as-tu': 247,
 'at': 85,
 'au': 8,
 'automne': 227,
 'automobile': 238,
 'aux': 15,
 'avez': 303,
 'avons': 353,
 'avril': 297,
 'banane': 159,
 'bananes': 219,
 'beau': 112,
 'belle': 224,
 'bien': 231,
 'blanc': 257,
 'blanche': 158,
 'bleu': 288,
 'bleue': 25,
 'brillant': 241,
 'brillante': 325,
 'bénigne': 88,
 "c'est": 260,
 'california': 103,
 'californie': 102,
 'calme': 180,
 'camion': 196,
 'ce': 72,
 'cet': 4,
 'cette': 87,
 'chat': 

### Check the Version of TensorFlow and Access to GPU
This will check to make sure you have the correct version of TensorFlow and access to a GPU

In [2]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
from distutils.version import LooseVersion
import warnings
import tensorflow as tf

# Check TensorFlow Version
assert LooseVersion(tf.__version__) in [LooseVersion('1.0.0'), LooseVersion('1.0.1')], 'This project requires TensorFlow version 1.0  You are using {}'.format(tf.__version__)
print('TensorFlow Version: {}'.format(tf.__version__))

# Check for a GPU
if not tf.test.gpu_device_name():
    warnings.warn('No GPU found. Please use a GPU to train your neural network.')
else:
    print('Default GPU Device: {}'.format(tf.test.gpu_device_name()))

TensorFlow Version: 1.0.0


//anaconda/lib/python3.6/site-packages/ipykernel_launcher.py:14: UserWarning: No GPU found. Please use a GPU to train your neural network.
  


## Build the Neural Network
You'll build the components necessary to build a Sequence-to-Sequence model by implementing the following functions below:
- `model_inputs`
- `process_decoding_input`
- `encoding_layer`
- `decoding_layer_train`
- `decoding_layer_infer`
- `decoding_layer`
- `seq2seq_model`

### Input
Implement the `model_inputs()` function to create TF Placeholders for the Neural Network. It should create the following placeholders:

- Input text placeholder named "input" using the TF Placeholder name parameter with rank 2.
- Targets placeholder with rank 2.
- Learning rate placeholder with rank 0.
- Keep probability placeholder named "keep_prob" using the TF Placeholder name parameter with rank 0.

Return the placeholders in the following the tuple (Input, Targets, Learing Rate, Keep Probability)

In [5]:
def model_inputs():
    """
    Create TF Placeholders for input, targets, and learning rate.
    :return: Tuple (input, targets, learning rate, keep probability)
    """
    # Implement Function
    Input = tf.placeholder(tf.int32, [None, None], name='input')
    Targets = tf.placeholder(tf.int32, [None, None])
    learning_rate = tf.placeholder(tf.float32)
    keep_prob = tf.placeholder(tf.float32, name='keep_prob')
    
    return Input, Targets, learning_rate, keep_prob

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_model_inputs(model_inputs)

Tests Passed


### Process Decoding Input
Implement `process_decoding_input` using TensorFlow to remove the last word id from each batch in `target_data` and concat the GO ID to the beginning of each batch.

In [6]:
def process_decoding_input(target_data, target_vocab_to_int, batch_size):
    """
    Preprocess target data for dencoding
    :param target_data: Target Placehoder
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param batch_size: Batch Size
    :return: Preprocessed target data
    """
    # Implement Function
    start = tf.fill([batch_size, 1], target_vocab_to_int['<GO>'])
    end = tf.strided_slice(target_data, [0,0], [batch_size,-1], [1,1])
    decoding_input = tf.concat([start, end], axis=1)

    return decoding_input

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_process_decoding_input(process_decoding_input)

Tests Passed


### Encoding
Implement `encoding_layer()` to create a Encoder RNN layer using [`tf.nn.dynamic_rnn()`](https://www.tensorflow.org/api_docs/python/tf/nn/dynamic_rnn).

In [7]:
def encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob):
    """
    Create encoding layer
    :param rnn_inputs: Inputs for the RNN
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param keep_prob: Dropout keep probability
    :return: RNN state
    """
    # Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    cell = tf.contrib.rnn.MultiRNNCell([drop]*num_layers)
    outputs, encoder_state = tf.nn.dynamic_rnn(cell, rnn_inputs, dtype=tf.float32)
    
    return encoder_state

"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_encoding_layer(encoding_layer)

Tests Passed


### Decoding - Training
Create training logits using [`tf.contrib.seq2seq.simple_decoder_fn_train()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_train) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder).  Apply the `output_fn` to the [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder) outputs.

In [8]:
def decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope,
                         output_fn, keep_prob):
    """
    Create a decoding layer for training
    :param encoder_state: Encoder State
    :param dec_cell: Decoder RNN Cell
    :param dec_embed_input: Decoder embedded input
    :param sequence_length: Sequence Length
    :param decoding_scope: TenorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Train Logits
    """
    # Implement Function
    dynamic_fn_train = tf.contrib.seq2seq.simple_decoder_fn_train(encoder_state)
    output_train, state_train, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(
        cell=dec_cell,\
        decoder_fn=dynamic_fn_train,\
        inputs=dec_embed_input,\
        sequence_length=sequence_length,\
        scope=decoding_scope)
    train_logits = output_fn(output_train)
    
    return train_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_train(decoding_layer_train)

Tests Passed


### Decoding - Inference
Create inference logits using [`tf.contrib.seq2seq.simple_decoder_fn_inference()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/simple_decoder_fn_inference) and [`tf.contrib.seq2seq.dynamic_rnn_decoder()`](https://www.tensorflow.org/versions/r1.0/api_docs/python/tf/contrib/seq2seq/dynamic_rnn_decoder). 

In [9]:
def decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id,
                         maximum_length, vocab_size, decoding_scope, output_fn, keep_prob):
    """
    Create a decoding layer for inference
    :param encoder_state: Encoder state
    :param dec_cell: Decoder RNN Cell
    :param dec_embeddings: Decoder embeddings
    :param start_of_sequence_id: GO ID
    :param end_of_sequence_id: EOS Id
    :param maximum_length: The maximum allowed time steps to decode
    :param vocab_size: Size of vocabulary
    :param decoding_scope: TensorFlow Variable Scope for decoding
    :param output_fn: Function to apply the output layer
    :param keep_prob: Dropout keep probability
    :return: Inference Logits
    """
    # Implement Function
    dynamic_fn_inference = tf.contrib.seq2seq.simple_decoder_fn_inference(
        output_fn, encoder_state, \
        dec_embeddings, start_of_sequence_id, \
        end_of_sequence_id, maximum_length, vocab_size)
    logits_infer, state_infer, _ = tf.contrib.seq2seq.dynamic_rnn_decoder(
        cell=dec_cell, \
        decoder_fn=dynamic_fn_inference, \
        scope=decoding_scope)
    
    return logits_infer


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer_infer(decoding_layer_infer)

Tests Passed


### Build the Decoding Layer
Implement `decoding_layer()` to create a Decoder RNN layer.

- Create RNN cell for decoding using `rnn_size` and `num_layers`.
- Create the output fuction using [`lambda`](https://docs.python.org/3/tutorial/controlflow.html#lambda-expressions) to transform it's input, logits, to class logits.
- Use the your `decoding_layer_train(encoder_state, dec_cell, dec_embed_input, sequence_length, decoding_scope, output_fn, keep_prob)` function to get the training logits.
- Use your `decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, start_of_sequence_id, end_of_sequence_id, maximum_length, vocab_size, decoding_scope, output_fn, keep_prob)` function to get the inference logits.

Note: You'll need to use [tf.variable_scope](https://www.tensorflow.org/api_docs/python/tf/variable_scope) to share variables between training and inference.

In [10]:
def decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size,
                   num_layers, target_vocab_to_int, keep_prob):
    """
    Create decoding layer
    :param dec_embed_input: Decoder embedded input
    :param dec_embeddings: Decoder embeddings
    :param encoder_state: The encoded state
    :param vocab_size: Size of vocabulary
    :param sequence_length: Sequence Length
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :param keep_prob: Dropout keep probability
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # Implement Function
    lstm = tf.contrib.rnn.BasicLSTMCell(rnn_size)
    drop = tf.contrib.rnn.DropoutWrapper(lstm, output_keep_prob=keep_prob)
    dec_cell = tf.contrib.rnn.MultiRNNCell([drop]*num_layers)
    
    with tf.variable_scope("decoding") as decoding_scope:
        output_fn = lambda x: tf.contrib.layers.fully_connected(x, vocab_size, activation_fn=None, scope=decoding_scope)
        
        #decoding_scope.reuse_variables()
        training_logits = decoding_layer_train(encoder_state, dec_cell, dec_embed_input,\
                                               sequence_length, decoding_scope, output_fn, \
                                               keep_prob)
        decoding_scope.reuse_variables()
        infer_logits = decoding_layer_infer(encoder_state, dec_cell, dec_embeddings, \
                                            target_vocab_to_int['<GO>'], \
                                            target_vocab_to_int['<EOS>'], \
                                            sequence_length, \
                                            vocab_size, decoding_scope, \
                                            output_fn, keep_prob)
    
    return training_logits, infer_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_decoding_layer(decoding_layer)

Tests Passed


### Build the Neural Network
Apply the functions you implemented above to:

- Apply embedding to the input data for the encoder.
- Encode the input using your `encoding_layer(rnn_inputs, rnn_size, num_layers, keep_prob)`.
- Process target data using your `process_decoding_input(target_data, target_vocab_to_int, batch_size)` function.
- Apply embedding to the target data for the decoder.
- Decode the encoded input using your `decoding_layer(dec_embed_input, dec_embeddings, encoder_state, vocab_size, sequence_length, rnn_size, num_layers, target_vocab_to_int, keep_prob)`.

In [11]:
def seq2seq_model(input_data, target_data, keep_prob, batch_size, sequence_length, source_vocab_size, target_vocab_size,
                  enc_embedding_size, dec_embedding_size, rnn_size, num_layers, target_vocab_to_int):
    """
    Build the Sequence-to-Sequence part of the neural network
    :param input_data: Input placeholder
    :param target_data: Target placeholder
    :param keep_prob: Dropout keep probability placeholder
    :param batch_size: Batch Size
    :param sequence_length: Sequence Length
    :param source_vocab_size: Source vocabulary size
    :param target_vocab_size: Target vocabulary size
    :param enc_embedding_size: Decoder embedding size
    :param dec_embedding_size: Encoder embedding size
    :param rnn_size: RNN Size
    :param num_layers: Number of layers
    :param target_vocab_to_int: Dictionary to go from the target words to an id
    :return: Tuple of (Training Logits, Inference Logits)
    """
    # Implement Function
    
    # Embedding to input data
    embedding_encoder = tf.Variable(tf.random_uniform((source_vocab_size, enc_embedding_size), -1, 1))
    enc_embed_input = tf.nn.embedding_lookup(embedding_encoder, input_data) #rnn_inputs
   
    # Encoder to input
    encoder_state = encoding_layer(enc_embed_input, rnn_size, num_layers, keep_prob)
    
    # Process target data
    decoding_input = process_decoding_input(target_data, target_vocab_to_int, batch_size)
    
    # Embedding to target data
    embedding_decoder = tf.Variable(tf.random_uniform((target_vocab_size, dec_embedding_size), -1, 1))
    dec_embed_input = tf.nn.embedding_lookup(embedding_decoder, decoding_input) 
    
    # Decode the encoded input
    training_logits, infer_logits = decoding_layer(dec_embed_input, embedding_decoder, encoder_state, \
                                                   target_vocab_size, sequence_length, rnn_size, num_layers, \
                                                   target_vocab_to_int, keep_prob)
    
    return training_logits, infer_logits


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_seq2seq_model(seq2seq_model)

Tests Passed


## Neural Network Training
### Hyperparameters
Tune the following parameters:

- Set `epochs` to the number of epochs.
- Set `batch_size` to the batch size.
- Set `rnn_size` to the size of the RNNs.
- Set `num_layers` to the number of layers.
- Set `encoding_embedding_size` to the size of the embedding for the encoder.
- Set `decoding_embedding_size` to the size of the embedding for the decoder.
- Set `learning_rate` to the learning rate.
- Set `keep_probability` to the Dropout keep probability

In [12]:
# Number of Epochs
epochs = 4
# Batch Size
batch_size = 256
# RNN Size
rnn_size = 256
# Number of Layers
num_layers = 3
# Embedding Size
encoding_embedding_size = 256
decoding_embedding_size = 256
# Learning Rate
learning_rate = 0.01
# Dropout Keep Probability
keep_probability = 0.6

### Build the Graph
Build the graph using the neural network you implemented.

In [13]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
save_path = 'checkpoints/dev'
(source_int_text, target_int_text), (source_vocab_to_int, target_vocab_to_int), _ = helper.load_preprocess()
max_source_sentence_length = max([len(sentence) for sentence in source_int_text])

train_graph = tf.Graph()
with train_graph.as_default():
    input_data, targets, lr, keep_prob = model_inputs()
    sequence_length = tf.placeholder_with_default(max_source_sentence_length, None, name='sequence_length')
    input_shape = tf.shape(input_data)
    
    train_logits, inference_logits = seq2seq_model(
        tf.reverse(input_data, [-1]), targets, keep_prob, batch_size, sequence_length, len(source_vocab_to_int), len(target_vocab_to_int),
        encoding_embedding_size, decoding_embedding_size, rnn_size, num_layers, target_vocab_to_int)

    tf.identity(inference_logits, 'logits')
    with tf.name_scope("optimization"):
        # Loss function
        cost = tf.contrib.seq2seq.sequence_loss(
            train_logits,
            targets,
            tf.ones([input_shape[0], sequence_length]))

        # Optimizer
        optimizer = tf.train.AdamOptimizer(lr)

        # Gradient Clipping
        gradients = optimizer.compute_gradients(cost)
        capped_gradients = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in gradients if grad is not None]
        train_op = optimizer.apply_gradients(capped_gradients)

### Train
Train the neural network on the preprocessed data. If you have a hard time getting a good loss, check the forms to see if anyone is having the same problem.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import time

def get_accuracy(target, logits):
    """
    Calculate accuracy
    """
    max_seq = max(target.shape[1], logits.shape[1])
    if max_seq - target.shape[1]:
        target = np.pad(
            target,
            [(0,0),(0,max_seq - target.shape[1])],
            'constant')
    if max_seq - logits.shape[1]:
        logits = np.pad(
            logits,
            [(0,0),(0,max_seq - logits.shape[1]), (0,0)],
            'constant')

    return np.mean(np.equal(target, np.argmax(logits, 2)))

train_source = source_int_text[batch_size:]
train_target = target_int_text[batch_size:]

valid_source = helper.pad_sentence_batch(source_int_text[:batch_size])
valid_target = helper.pad_sentence_batch(target_int_text[:batch_size])

with tf.Session(graph=train_graph) as sess:
    sess.run(tf.global_variables_initializer())
    
    loss_list = []
    valid_acc_list = []
    for epoch_i in range(epochs):
        for batch_i, (source_batch, target_batch) in enumerate(
                helper.batch_data(train_source, train_target, batch_size)):
            start_time = time.time()
            
            _, loss = sess.run(
                [train_op, cost],
                {input_data: source_batch,
                 targets: target_batch,
                 lr: learning_rate,
                 sequence_length: target_batch.shape[1],
                 keep_prob: keep_probability})
            
            batch_train_logits = sess.run(
                inference_logits,
                {input_data: source_batch, keep_prob: 1.0})
            batch_valid_logits = sess.run(
                inference_logits,
                {input_data: valid_source, keep_prob: 1.0})
                
            train_acc = get_accuracy(target_batch, batch_train_logits)
            valid_acc = get_accuracy(np.array(valid_target), batch_valid_logits)
            end_time = time.time()
            
            loss_list.append(loss)
            valid_acc_list.append(valid_acc)
            print('Epoch {:>3} Batch {:>4}/{} - Train Accuracy: {:>6.3f}, Validation Accuracy: {:>6.3f}, Loss: {:>6.3f}'
                  .format(epoch_i, batch_i, len(source_int_text) // batch_size, train_acc, valid_acc, loss))

    # Save Model
    saver = tf.train.Saver()
    saver.save(sess, save_path)
    print('Model Trained and Saved')
   
    # Visualize the loss and accuracy
    import matplotlib.pyplot as plt
    f, (ax1, ax2) = plt.subplots(1, 2, figsize=(18, 6))
    ax1.plot(loss_list, color='red')
    ax1.set_title('Traning Loss')

    ax1.set_ylabel('Loss value')
    ax2.plot(valid_acc_list)
    ax2.set_xlabel('Iterations')
    ax2.set_ylabel('Accuracy')
    ax2.set_title('Validation Accuracy')
    plt.show()

Epoch   0 Batch    0/538 - Train Accuracy:  0.243, Validation Accuracy:  0.323, Loss:  5.876
Epoch   0 Batch    1/538 - Train Accuracy:  0.231, Validation Accuracy:  0.316, Loss:  4.577
Epoch   0 Batch    2/538 - Train Accuracy:  0.275, Validation Accuracy:  0.336, Loss:  5.211
Epoch   0 Batch    3/538 - Train Accuracy:  0.238, Validation Accuracy:  0.317, Loss:  3.884
Epoch   0 Batch    4/538 - Train Accuracy:  0.272, Validation Accuracy:  0.349, Loss:  3.808
Epoch   0 Batch    5/538 - Train Accuracy:  0.308, Validation Accuracy:  0.355, Loss:  3.696
Epoch   0 Batch    6/538 - Train Accuracy:  0.305, Validation Accuracy:  0.347, Loss:  3.569
Epoch   0 Batch    7/538 - Train Accuracy:  0.288, Validation Accuracy:  0.355, Loss:  3.548
Epoch   0 Batch    8/538 - Train Accuracy:  0.283, Validation Accuracy:  0.351, Loss:  3.468
Epoch   0 Batch    9/538 - Train Accuracy:  0.283, Validation Accuracy:  0.351, Loss:  3.409
Epoch   0 Batch   10/538 - Train Accuracy:  0.264, Validation Accuracy

Epoch   0 Batch   89/538 - Train Accuracy:  0.453, Validation Accuracy:  0.481, Loss:  0.967
Epoch   0 Batch   90/538 - Train Accuracy:  0.462, Validation Accuracy:  0.485, Loss:  0.936
Epoch   0 Batch   91/538 - Train Accuracy:  0.424, Validation Accuracy:  0.482, Loss:  0.957
Epoch   0 Batch   92/538 - Train Accuracy:  0.437, Validation Accuracy:  0.477, Loss:  0.949
Epoch   0 Batch   93/538 - Train Accuracy:  0.417, Validation Accuracy:  0.474, Loss:  0.950
Epoch   0 Batch   94/538 - Train Accuracy:  0.418, Validation Accuracy:  0.487, Loss:  0.938
Epoch   0 Batch   95/538 - Train Accuracy:  0.485, Validation Accuracy:  0.475, Loss:  0.848
Epoch   0 Batch   96/538 - Train Accuracy:  0.460, Validation Accuracy:  0.470, Loss:  0.870
Epoch   0 Batch   97/538 - Train Accuracy:  0.412, Validation Accuracy:  0.468, Loss:  0.928
Epoch   0 Batch   98/538 - Train Accuracy:  0.470, Validation Accuracy:  0.482, Loss:  0.863
Epoch   0 Batch   99/538 - Train Accuracy:  0.421, Validation Accuracy

Epoch   0 Batch  178/538 - Train Accuracy:  0.593, Validation Accuracy:  0.587, Loss:  0.646
Epoch   0 Batch  179/538 - Train Accuracy:  0.559, Validation Accuracy:  0.572, Loss:  0.683
Epoch   0 Batch  180/538 - Train Accuracy:  0.591, Validation Accuracy:  0.572, Loss:  0.645
Epoch   0 Batch  181/538 - Train Accuracy:  0.536, Validation Accuracy:  0.571, Loss:  0.695
Epoch   0 Batch  182/538 - Train Accuracy:  0.547, Validation Accuracy:  0.569, Loss:  0.673
Epoch   0 Batch  183/538 - Train Accuracy:  0.587, Validation Accuracy:  0.573, Loss:  0.622
Epoch   0 Batch  184/538 - Train Accuracy:  0.582, Validation Accuracy:  0.568, Loss:  0.619
Epoch   0 Batch  185/538 - Train Accuracy:  0.546, Validation Accuracy:  0.571, Loss:  0.658
Epoch   0 Batch  186/538 - Train Accuracy:  0.556, Validation Accuracy:  0.571, Loss:  0.647
Epoch   0 Batch  187/538 - Train Accuracy:  0.594, Validation Accuracy:  0.579, Loss:  0.614
Epoch   0 Batch  188/538 - Train Accuracy:  0.565, Validation Accuracy

Epoch   0 Batch  267/538 - Train Accuracy:  0.605, Validation Accuracy:  0.645, Loss:  0.550
Epoch   0 Batch  268/538 - Train Accuracy:  0.643, Validation Accuracy:  0.646, Loss:  0.525
Epoch   0 Batch  269/538 - Train Accuracy:  0.616, Validation Accuracy:  0.646, Loss:  0.550
Epoch   0 Batch  270/538 - Train Accuracy:  0.624, Validation Accuracy:  0.647, Loss:  0.558
Epoch   0 Batch  271/538 - Train Accuracy:  0.628, Validation Accuracy:  0.647, Loss:  0.556
Epoch   0 Batch  272/538 - Train Accuracy:  0.607, Validation Accuracy:  0.648, Loss:  0.579
Epoch   0 Batch  273/538 - Train Accuracy:  0.629, Validation Accuracy:  0.646, Loss:  0.557
Epoch   0 Batch  274/538 - Train Accuracy:  0.601, Validation Accuracy:  0.646, Loss:  0.587
Epoch   0 Batch  275/538 - Train Accuracy:  0.613, Validation Accuracy:  0.645, Loss:  0.571
Epoch   0 Batch  276/538 - Train Accuracy:  0.643, Validation Accuracy:  0.655, Loss:  0.558
Epoch   0 Batch  277/538 - Train Accuracy:  0.635, Validation Accuracy

Epoch   0 Batch  356/538 - Train Accuracy:  0.657, Validation Accuracy:  0.661, Loss:  0.481
Epoch   0 Batch  357/538 - Train Accuracy:  0.674, Validation Accuracy:  0.658, Loss:  0.503
Epoch   0 Batch  358/538 - Train Accuracy:  0.639, Validation Accuracy:  0.660, Loss:  0.513
Epoch   0 Batch  359/538 - Train Accuracy:  0.658, Validation Accuracy:  0.659, Loss:  0.496
Epoch   0 Batch  360/538 - Train Accuracy:  0.657, Validation Accuracy:  0.658, Loss:  0.510
Epoch   0 Batch  361/538 - Train Accuracy:  0.657, Validation Accuracy:  0.656, Loss:  0.489
Epoch   0 Batch  362/538 - Train Accuracy:  0.669, Validation Accuracy:  0.652, Loss:  0.480
Epoch   0 Batch  363/538 - Train Accuracy:  0.648, Validation Accuracy:  0.652, Loss:  0.475
Epoch   0 Batch  364/538 - Train Accuracy:  0.618, Validation Accuracy:  0.651, Loss:  0.528
Epoch   0 Batch  365/538 - Train Accuracy:  0.612, Validation Accuracy:  0.622, Loss:  0.490
Epoch   0 Batch  366/538 - Train Accuracy:  0.644, Validation Accuracy

### Save Parameters
Save the `batch_size` and `save_path` parameters for inference.

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
# Save parameters for checkpoint
helper.save_params(save_path)

# Checkpoint

In [ ]:
"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
import tensorflow as tf
import numpy as np
import helper
import problem_unittests as tests

_, (source_vocab_to_int, target_vocab_to_int), (source_int_to_vocab, target_int_to_vocab) = helper.load_preprocess()
load_path = helper.load_params()

## Sentence to Sequence
To feed a sentence into the model for translation, you first need to preprocess it.  Implement the function `sentence_to_seq()` to preprocess new sentences.

- Convert the sentence to lowercase
- Convert words into ids using `vocab_to_int`
 - Convert words not in the vocabulary, to the `<UNK>` word id.

In [ ]:
def sentence_to_seq(sentence, vocab_to_int):
    """
    Convert a sentence to a sequence of ids
    :param sentence: String
    :param vocab_to_int: Dictionary to go from the words to an id
    :return: List of word ids
    """
    # TODO: Implement Function
    return None


"""
DON'T MODIFY ANYTHING IN THIS CELL THAT IS BELOW THIS LINE
"""
tests.test_sentence_to_seq(sentence_to_seq)

## Translate
This will translate `translate_sentence` from English to French.

In [ ]:
translate_sentence = 'he saw a old yellow truck .'


"""
DON'T MODIFY ANYTHING IN THIS CELL
"""
translate_sentence = sentence_to_seq(translate_sentence, source_vocab_to_int)

loaded_graph = tf.Graph()
with tf.Session(graph=loaded_graph) as sess:
    # Load saved model
    loader = tf.train.import_meta_graph(load_path + '.meta')
    loader.restore(sess, load_path)

    input_data = loaded_graph.get_tensor_by_name('input:0')
    logits = loaded_graph.get_tensor_by_name('logits:0')
    keep_prob = loaded_graph.get_tensor_by_name('keep_prob:0')

    translate_logits = sess.run(logits, {input_data: [translate_sentence], keep_prob: 1.0})[0]

print('Input')
print('  Word Ids:      {}'.format([i for i in translate_sentence]))
print('  English Words: {}'.format([source_int_to_vocab[i] for i in translate_sentence]))

print('\nPrediction')
print('  Word Ids:      {}'.format([i for i in np.argmax(translate_logits, 1)]))
print('  French Words: {}'.format([target_int_to_vocab[i] for i in np.argmax(translate_logits, 1)]))

## Imperfect Translation
You might notice that some sentences translate better than others.  Since the dataset you're using only has a vocabulary of 227 English words of the thousands that you use, you're only going to see good results using these words.  For this project, you don't need a perfect translation. However, if you want to create a better translation model, you'll need better data.

You can train on the [WMT10 French-English corpus](http://www.statmt.org/wmt10/training-giga-fren.tar).  This dataset has more vocabulary and richer in topics discussed.  However, this will take you days to train, so make sure you've a GPU and the neural network is performing well on dataset we provided.  Just make sure you play with the WMT10 corpus after you've submitted this project.
## Submitting This Project
When submitting this project, make sure to run all the cells before saving the notebook. Save the notebook file as "dlnd_language_translation.ipynb" and save it as a HTML file under "File" -> "Download as". Include the "helper.py" and "problem_unittests.py" files in your submission.